docs/JAX_porting_PyTorch_model.md

```diff
@@ -0,0 +1,1607 @@
+```
+
+A MaxViT layer is composed of: MBConv, window_attention as `PartitionAttentionLayer` and grid_attention as `PartitionAttentionLayer`.
+
+```{code-cell} ipython3
+[[mod.__class__.__name__ for mod in maxvit_layer.layers] for b in torch_model.blocks for maxvit_layer in b.layers]
+```
+
+### Inference on data
+
+Let's check the model on dummy input and on a real image
+
+```{code-cell} ipython3
+import torch
+
+
+torch_model.eval()
+with torch.inference_mode():
+    x = torch.rand(2, 3, 224, 224)
+    output = torch_model(x)
+
+print(output.shape)
+```
+
+We can download an image of a Pembroke Corgy dog from TorchVision's gallery ([url](https://github.com/pytorch/vision/blob/main/gallery/assets/dog1.jpg?raw=true)) together with ImageNet classes dictionary ([url](https://raw.githubusercontent.com/pytorch/vision/refs/heads/main/gallery/assets/imagenet_class_index.json)):
+
+```bash
+wget "https://github.com/pytorch/vision/blob/main/gallery/assets/dog1.jpg?raw=true" -O dog1.jpg
+wget "https://raw.githubusercontent.com/pytorch/vision/refs/heads/main/gallery/assets/imagenet_class_index.json" -O imagenet_class_index.json
+```
+
+```{code-cell} ipython3
+import matplotlib.pyplot as plt
+
+import json
+from torchvision.io import read_image
+
+
+preprocess = MaxVit_T_Weights.IMAGENET1K_V1.transforms()
+
+with open("imagenet_class_index.json") as labels_file:
+  labels = json.load(labels_file)
+
+
+dog1 = read_image("dog1.jpg")
+tensor = preprocess(dog1)
+
+torch_model.eval()
+with torch.inference_mode():
+  output = torch_model(tensor.unsqueeze(dim=0))
+
+class_id = output.argmax(dim=1).item()
+
+print(f"Prediction for the Dog: {labels[str(class_id)]}, score: {output.softmax(dim=-1)[0, class_id]}")
+
+plt.title(f"{labels[str(class_id)]}\nScore: {output.softmax(dim=-1)[0, class_id]}")
+plt.imshow(dog1.permute(1, 2, 0))
+```
+
+## Port MaxViT model to JAX
+
+Let us now define all the necessary model parts in JAX using Flax module. PyTorch model's implementation can be found [here](https://pytorch.org/vision/stable/_modules/torchvision/models/maxvit.html#maxvit_t), we can now list the modules and helper functions we implement in JAX:
+
+- Conv2dNormActivation
+- SqueezeExcitation
+- StochasticDepth
+- MBConv
+- RelativePositionalMultiHeadAttention
+- SwapAxes
+- WindowPartition
+- WindowDepartition
+- PartitionAttentionLayer
+- MaxVitLayer
+- MaxVitBlock
+- MaxViT
+
+Flax NNX module is very similar to PyTorch `torch.nn` module and we can simply map the following modules between PyTorch and Flax:
+- `nn.Sequential` and `nn.ModuleList` -> `nnx.Sequential`
+- `nn.Linear` -> `nnx.Linear`
+- `nn.Conv2d` -> `nnx.Conv`
+- `nn.BatchNorm2d` -> `nnx.BatchNorm`
+- Activations like `nn.ReLU` -> `lambda x: nnx.relu(x)`
+- Pooling layers like `nn.AvgPool2d` -> `lambda x: nnx.avg_pool(x, ...)`
+- `nn.AdaptiveAvgPool2d(1)` -> `lambda x: nnx.avg_pool(x, (x.shape[1], x.shape[2]))`, x is in NHWC format
+- `nn.Flatten()` -> `lambda x: x.reshape(x.shape[0], -1)`
```

Comment:
maybe use `jax.vmap(jnp.ravel)` here? Or is that too magic?

Reply:
yes, this is too magic.